pystiche_papers/ulyanov_et_al_2016/_utils.py

```diff
@@ -34,9 +34,9 @@ def hyper_parameters(
             layers=("relu1_1", "relu2_1", "relu3_1", "relu4_1")
             if impl_params and instance_norm
             else ("relu1_1", "relu2_1", "relu3_1", "relu4_1", "relu5_1"),
-            layer_weights="sum",
+            layer_weights=[1e3] * 4 if impl_params and instance_norm else [1e3] * 5,
             # https://github.com/pmeier/texture_nets/blob/b2097eccaec699039038970b191780f97c238816/stylization_train.lua#L23
-            score_weight=1e3 if impl_params and not instance_norm else 1e0,
+            score_weight=1e0 if impl_params and not instance_norm else 1e0,
```

Comment:
The `score_weight` could actually be completely omitted here, since `score_weight=1e0` is the default value. 

Something also went wrong when transferring the correct `layer_weights`,  as this should be something like:
```python
layer_weights=1e3 if impl_params and not instance_norm else 1e0,
```
Since the `score_weight` is `1e3` in only one of the reference implementations.
 So I would suggest to use the code line above and set the length of the required Sequence[float] of the `MultiLayerEncodingOperator` in `style_loss`. Or what do you think?

Reply:
I'm sorry, but I don't understand your comment. Maybe there is some confusion about what `score_weight` and `layer_weights` are. Basically, each layer is weighted with `score_weight * layer_weights[i]` where `i` denotes the layer index. Thus

- `score_weight=1e3` and `layer_weights="sum"` (and thus `layer_weights=[1.0] * len(layers)`) and
- `score_weight=1e0` and `layer_weights=[1e3] * len(layers)`

are equal. The former has to advantage that it works for any number of layers whereas you need to specify it in the second case. Furthermore the former is conceptually more clear: Instead of the individual layers, the style loss is weighted. 

If that does not help could you please compile an example what you want to achieve with this?

Reply:
I'm completely with you on the different weights. However, this will no longer work with the manipulated backward pass into the `GramOperator`, since the `score_weight`:
```python
score_weight=1e3 if impl_params and not instance_norm else 1e0
```
for each `GramOperator` is required here. If the `layer_weights` are `"sum"` the `score_weight` of these `GramOperator`s will always be `1e0`. This was also my intention why I changed the `layer_weights` here, because this way the operators will have the needed `score_weight`.

Reply:
I think I finally got it: I think you are saying that this is problematic since the backward pass of `ManipulateGradient` is **not** independent of `score_weight`. In the reference implementation the `score_weight` is `1e0` and the `layer_weight` is `1e3`, correct?

I've checked and indeed this is a problem. Good catch if you intended this and otherwise thank you for sending me on that journey :wink:

---

Here is what I would do about that:

You put 

```python
# https://github.com/pmeier/texture_nets/blob/aad2cc6f8a998fedc77b64bdcfe1e2884aa0fb3e/train.lua#L44
style_loss_layers = (
    ("relu1_1", "relu2_1", "relu3_1", "relu4_1")
    if impl_params and instance_norm
    else ("relu1_1", "relu2_1", "relu3_1", "relu4_1", "relu5_1")
)
```

on the top of the function and within the hyper parameters definition you can do 

```python
style_loss = HyperParameters(
    layers=style_loss_layers,
    # https://github.com/pmeier/texture_nets/blob/b2097eccaec699039038970b191780f97c238816/stylization_train.lua#L23
    layer_weights=(
        [1e3 if impl_params and not instance_norm else 1e0] * len(style_loss_layers)
    ),
    score_weight=1e0,
)
```

You should also include a short comment here why we need to go with this approach rather than the usual one.